3.next - Add deprecation warnings for public request properties

markstory opened the pull request:
Having public properties on the request is not something we can have in a PSR7 style request object. I've added deprecation warnings to these properties by using magic methods. This may result in small issues, but we don't have many other tools to deprecate public properties.

Unfortunately it is another very large pull request :cry:. This should be the last one though.

Refs #11385

markstory committed:
Deprecate public properties on ServerRequest.

Also fix the tests for that class. There are many more tests that need
to be updated still.

markstory committed:
Fix deprecated property usage in routing.

markstory committed:
Fix SecurityComponent using deprecated properties.

This has required a minor breaking change in `generateToken()`. Because
of how this method was setup it was impossible to make a backwards
compatible change. I'm relying on the hope that not many people directly
invoke this method and that the break will not disrupt people too
greatly.

markstory committed:
Stop using deprecated properties in Components.

Update components and their tests to not use deprecated properties.

markstory committed:
Let people know how to replace all the request data more easily.

markstory committed:
Update controller test.

markstory committed:
Fix Error tests using deprecated request properties.

markstory committed:
Fix request usage in Auth tests.

markstory committed:
Fix deprecated property use in View test.

markstory committed:
Fix more tests using deprecated request features.

markstory committed:
Fix deprecated request property use in paginator helper & tests.

markstory committed:
Fix deprecated request property use in HtmlHelper tests.

markstory committed:
Fix some of formhelper's tests.

markstory committed:
Fix warnings on remaining FormHelper tests.

markstory committed:
Fix remaining failing helper tests.

markstory committed:
Fix json view test.

markstory committed:
Fix more deprecated feature use.

markstory committed:
Fix lint errors.

markstory committed:
Remove unused objects.

markstory committed:
Fix annotations point to Network\Session

markstory committed:
Fix up phpstan errors.

markstory committed:
fix phpcs

saeideng commented on src/Routing/Router.php:
```diff
@@ -902,17 +902,15 @@ public static function parseNamedParams(ServerRequest $request, array $options =
-
-            return $request;
+        if (!$request->getParam('pass')) {
+            return $request->withParam('named', []);
```
```‍‍‍php
$request->params['named'] = [];
return $request;
```
is not equal to 
```php
return $request->withParam('named', []);
```
i mean 
```php
return $request=$request->withParam('named', []);
```

markstory replied:
How are they not equivalent? In both cases the mutated request is returned.

markstory replied:
The difference is the request is not mutated in place. I've updated the tests to use the return value and will update the migration guide as well.

saeideng commented on src/Routing/Router.php:
```diff
@@ -933,9 +931,7 @@ public static function parseNamedParams(ServerRequest $request, array $options =
-        $request->params['named'] = $named;
-
-        return $request;
+        return $request->withParam('named', $named);
```
and here

markstory committed:
Fix failing tests.

requests are now treated as immutable internally. If you are using
Router::parseNamedParams() you must capture the return value.
